content/blog/2021/2021-04-25-joking-for-the-win-seriously.md

```diff
@@ -0,0 +1,56 @@
+```
+learn targetdir trainingdata epochs 3000 batchsize 1000000 reduction_gameply 12 smart_fen_skipping use_draw_in_training 1 use_draw_in_validation 1 lr 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 1 skip_duplicated_positions_in_training 0 eval_save_interval 250000000 loss_output_interval 1000000 validation_set_file_name validationdata\val.binpack validation_count 10000
+```
+I had to restart 2-3 times after a few steps as it was clear the random initialization wasn't working. I then repeated the process with a 0.1 learning rate starting from the final net generated, and once more with 0.01 learning rate. Training at each learning rate took a bit more than 7 hours, and was done over a weekend.  I’m not claiming this is a good training procedure as I have no real experience with NNUE training and several of the values were just guesses. I expect people who know what they are doing can generate a better net from the same training data, and would be very happy to see the results.
+
+To summarize: 
+
+- The first positive outcome from this joke was the addition of code to the rescorer to output NNUE plain format (for details see below).
+- Moreover, the lcfish training data were contributed to the [Stockfish training data collection](https://drive.google.com/drive/folders/1GhSF8s-ljJlsaPibtEgn1AQ6aGZanIGz?usp=sharing).
+- We also have the single line change to allow building lc0 on windows with gcc (but not all backends).
+- Finally, some of the training data generated crashed the Stockfish-NNUE trainer, which led to the discovery and fix of a 3 year old bug in lc0.
+
+Not bad for an April Fools’ joke!
+
+**PS.** To convert Leela training data with the rescorer the `--nnue-plain-file=output_file.plain`option is used. Note that the data are appended to the .plain file specified if it is there, and that the input files are deleted unless `--delete-files=false` is also specified. So a typical command would be
+```
+./rescorer rescore --input=/path/to/training/data --syzygy-paths=/path/to/syzygy/files --nnue-plain-file=output_file.plain
+```
+(the rescorer requirees syzygy files to be available).
+
+The current training data (V6 format)t store both the played move, but also the best move, which for training games is very often different, as well as the respective evaluations. This gives us the option to save either the best move evaluation (default) or the played move one (`--nnue-best-score=false`). Similarly the generated file can contain the played move for each position (default) of the best move (`--nnue-best-move=true`), but note that changing this will typically write a .plain file that doesn’t compress very well when converted to a .binpack one.
```

Comment:
spurious t

Reply:
of the best move --> or

Reply:
maybe add explanation for the last sentence (played move gives next FEN), as it is an interesting detail.

Reply:
agreed